USF-1670: Dropins Localization (#95)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/all/localizing.mdx

---
title: Localizing dropins

> description: How to localize your dropins.

Reviewer: ```suggestion
description: Learn how to localize your drop-in components.
```

---

import Aside from '@components/Aside.astro';
import Diagram from '@components/Diagram.astro';
import CodeImport from '@components/CodeImport.astro';
import CodeInclude from '@components/CodeInclude.astro';
import dictionary from '@dropins/tools/types/elsie/src/i18n/en_US.json.d.ts?raw';
import Vocabulary from '@components/Vocabulary.astro';

In our global economy, localization is crucial. And it's one of the easiest changes to make. This page shows how to add a second language file and use it in your storefront.

## Big picture

> The following diagram shows the process for adding and overriding labels and text for drop-ins within the boilerplate template.

Reviewer: ```suggestion
The following diagram shows the process for adding and overriding labels and text for drop-in components within the boilerplate template.
```

<Diagram caption="How localization and labeling works in storefronts.">
  ![How localization and labeling works in storefronts.](@images/DropinDictionaries.svg)
</Diagram>

## Vocabulary

<Vocabulary>

> ### Dropin dictionaries

Reviewer: ```suggestion
### Dictionaries
```

> JSON files that defines the shape, keys, and default values for labels and text within a drop-in. Each drop-in has a default dictionary file: `en_US.json`. But drop-ins can contain multiple dictionary files as the site is localized for other languages, including `fr_CA.json` for French, `de_DE.json` for German, and so on.

Reviewer: ```suggestion
JSON files that define the shape, keys, and default values for labels and text within a drop-in component. Each drop-in component has a default dictionary file: `en_US.json`. But drop-in components can contain multiple dictionary files as the site is localized for other languages, including `fr_CA.json` for French, `de_DE.json` for German, and so on.
```

### Language code

A two-letter code that identifies a language. For example, `en` for English, `fr` for French, and `de` for German.

### Language files

> See Dropin dictionaries.

Reviewer: ```suggestion
See dictionaries.
```

### Locale

A combination of a language code and a country code. For example, `en_US` for English in the United States, `fr_CA` for French in Canada, and `de_DE` for German in Germany.

### Localization

The process of adapting a product or service to a particular language.

### Placeholders

> A Google Sheets file (`placeholders`) or a SharePoint Excel file (`placeholders.xlsx`) where you enter your site's text variables, which includes localization. The file has two columns: a _Key column_ that contains the variables in your site's code and a _Text column_ that contains the text values that will replace those variables at runtime. See https://www.aem.live/docs/placeholders for more information.

Reviewer: ```suggestion
A Google Sheets file (`placeholders`) or a SharePoint Excel file (`placeholders.xlsx`) where you enter your site's text variables, which includes localization. The file has two columns: a _Key column_ that contains the variables in your site's code and a _Text column_ that contains the text values that will replace those variables at runtime. See [placeholders](https://www.aem.live/docs/placeholders) in the Edge Delivery Services documentation for more information.
```

</Vocabulary>

> ## Common dropin dictionary

Reviewer: ```suggestion
## Common drop-in component dictionary
```

> The following code block shows the default dictionary file for all drop-ins. This file contains the default values for all labels and text common to all drop-ins.

Reviewer: ```suggestion
The following code block shows the default dictionary file for all drop-in components. This file contains the default values for all labels and text common to all drop-in components.
```

> <CodeImport code={dictionary} title="Common Dropin Dictionary" lang="typescript" frame="none" />

Reviewer: ```suggestion
<CodeImport code={dictionary} title="Common drop-in component dictionary" lang="typescript" frame="none" />
```

> ## How to override drop-in dictionaries

Reviewer: ```suggestion
## How to override drop-in component dictionaries
```

> To override the default dictionary values, you can create a new dictionary in the boilerplate's `product-details.js` file and assign it to the drop-in's `langDefinitions` constant. The new dictionary file should contain only the keys and values you want to override in the default dictionary.

Reviewer: ```suggestion
To override the default dictionary values, you can create a new dictionary in the boilerplate's `product-details.js` file and assign it to the drop-in component's `langDefinitions` constant. The new dictionary file should contain only the keys and values you want to override in the default dictionary.
```

## Big picture

The following diagram shows the process for adding and overriding labels and text for drop-in components within the boilerplate template.

<Diagram caption="How localization and labeling works in storefronts.">
  ![How localization and labeling works in storefronts.](@images/DropinDictionaries.svg)
</Diagram>

## Vocabulary

<Vocabulary>

### Dictionaries

JSON files that define the shape, keys, and default values for labels and text within a drop-in component. Each drop-in component has a default dictionary file: `en_US.json`. But drop-in components can contain multiple dictionary files as the site is localized for other languages, including `fr_CA.json` for French, `de_DE.json` for German, and so on.

### Language code

A two-letter code that identifies a language. For example, `en` for English, `fr` for French, and `de` for German.

### Language files

See dictionaries.

### Locale

A combination of a language code and a country code. For example, `en_US` for English in the United States, `fr_CA` for French in Canada, and `de_DE` for German in Germany.

### Localization

The process of adapting a product or service to a particular language.

### Placeholders

> A Google Sheets file (`placeholders`) or a SharePoint Excel file (`placeholders.xlsx`) where you enter your site's text variables, which includes localization. The file has two columns: a _Key column_ that contains the variables in your site's code and a _Text column_ that contains the text values that will replace those variables at runtime. See [placeholders](https://www.aem.live/docs/placeholders) in the Edge Delivery Services documentation for more information.

Reviewer 2: Should we talk about key format? @fnhipster recently did some work there in the boilerplate and I think he also may have found some inconsistencies between how we define placeholder keys and how EDS defines them.

Author: I updated the steps based on the latest added by @fnhipster to the `product-details.js` block in the `main` branch.

</Vocabulary>

## Common drop-in component dictionary

The following code block shows the default dictionary file for all drop-in components. This file contains the default values for all labels and text common to all drop-in components.

> <CodeImport code={dictionary} title="Common drop-in component dictionary" lang="typescript" frame="none" />

Reviewer 2: I think this may be potentially confusing.

First, there is this core/common dictionary that is used to provide some basic text for components from the SDK, but is there a guarantee that our dropins will use _those_ without modification? Or can they overwrite with some other value. I'm not sure how often this happens in practice but I believe it is a possibility.

Beyond that, I think it's mainly confusing around the word "components" as I have heard of things like the Checkout Dropin and the Cart Dropin be referred to as "dropin components".

Author: I removed this section in the latest push.
